src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,229 @@
+use diesel::{
+    dsl::now, BoolExpressionMethods, Connection, ExpressionMethods, NullableExpressionMethods,
+    PgConnection, QueryDsl, QueryResult, RunQueryDsl, SelectableHelper,
+};
+use std::sync::Arc;
+
+/// The threshold in days for the expiry notification.
+const EXPIRY_THRESHOLD: chrono::TimeDelta = chrono::TimeDelta::days(3);
+
+/// The maximum number of tokens to check.
+/// To avoid sending too many emails and submitting to many transactions, we limit the number of
+/// tokens to check.
+const MAX_ROWS: i64 = 10000;
+/// A job responsible for monitoring the status of a token.
+/// It checks if the token is about to reach its expiry date.
+/// If the token is about to expire, the job triggers a notification.
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct SendTokenExpiryNotifications;
+
+impl BackgroundJob for SendTokenExpiryNotifications {
+    const JOB_NAME: &'static str = "expiry_notification";
+
+    type Context = Arc<Environment>;
+
+    #[instrument(skip(env), err)]
+    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
+        let conn = env.deadpool.get().await?;
+        conn.interact(move |conn| {
+            // Check if the token is about to expire
+            // If the token is about to expire, trigger a notification.
+            check(&env.emails, conn)
+        })
+        .await
+        .map_err(|err| anyhow!(err.to_string()))?
+    }
+}
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens = find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    if expired_tokens.len() == MAX_ROWS as usize {
+        warn!("The number of tokens to check exceeds the limit");
+    }
+    for token in &expired_tokens {
+        conn.transaction(|conn| {
+            // Send notification.
+            let user = User::find(conn, token.user_id)?;
+            let Some(recipient) = user.email(conn)? else {
+                return Err(anyhow!("No address found"));
+            };
+            let email = ExpiryNotificationEmail {
+                name: &user.gh_login,
+                token_name: &token.name,
+                expiry_date: token.expired_at.unwrap().and_utc(),
+            };
+            match emails.send(&recipient, email) {
+                Ok(_) => {
+                    // Update the token to prevent duplicate notifications.
+                    diesel::update(token)
+                        .set(api_tokens::expiry_notification_at.eq(now.nullable()))
+                        .execute(conn)?;
+                }
+                Err(e) => {
+                    error!(?e, ?recipient, "Failed to send notification");
+                    return Err(anyhow!("Failed to send notification: {}", e));
+                }
+            }
+            Ok::<_, anyhow::Error>(())
+        })?;
```

Comment:
if we use `?` here then the loop will stop on the first error, but I guess we would probably want to continue processing the remaining tokens?